RBAC Docs (#2486)
pulp/pulp_rpm: docs/rbac.rst

    pulp role show --name "rpm.rpmremote_viewer"

.. code-block:: shell

    {
      "pulp_href": "/pulp/api/v3/roles/62baf888-09c2-4c18-8d3a-354d9c6bdb48/",
      "pulp_created": "2022-04-07T12:41:11.065567Z",
      "name": "rpm.rpmremote_viewer",
      "description": null,
      "permissions": [
        "rpm.view_rpmremote"
      ],
      "locked": true
    }

Creator Role
^^^^^^^^^^^^

This role contains only ``add`` permission for a specific endpoint to allow a user to create a new object of its type.
Be aware that you may need additional permission to create a new object.
For example, ``rpm.rpmrepository_view`` is needed for a used repository to create a new publication.

.. code-block:: shell
   :caption: Example of the requested condition to create an RPM Publication

    {
        "action": ["create"],
        "principal": "authenticated",
        "effect": "allow",
        "condition": [
            "has_model_perms:rpm.add_rpmpublication",
            "has_repo_attr_model_or_obj_perms:rpm.view_rpmrepository",
        ]
    }

Owner Role
^^^^^^^^^^

This role contains all permissions needed for manipulating objects, such as modifying or removing them.

> This role also contains permission to manage permission on the object, for example, adding or removing permissions.

Reviewer: ```suggestion
This role also contains permission to manage role assignments on the object, for example, adding or removing user roles.
```

By default, the user who created a new object will receive this role.

.. code-block:: shell
   :caption: Example of the RPM Remote owner role

    {
      "pulp_href": "/pulp/api/v3/roles/592356da-8cdf-4a1b-a434-3f10cda285f7/",
      "pulp_created": "2022-04-07T12:41:11.047168Z",
      "name": "rpm.rpmremote_owner",
      "description": null,
      "permissions": [
        "rpm.change_rpmremote",
        "rpm.delete_rpmremote",
        "rpm.manage_roles_rpmremote",
        "rpm.view_rpmremote"
      ],
      "locked": true
    }

RPM specific Roles
------------------

Also, the RPM plugin ships two roles to use.

RPM Admin Role (rpm.admin)
^^^^^^^^^^^^^^^^^^^^^^^^^^

This role is like a pulp admin but only for the RPM plugin.

> You can do every RPM workflow. If you need a more

Reviewer: ```suggestion
You can perform every RPM workflow. If you need a more
```

         "rpm.add_ulnremote",
         "rpm.change_rpmalternatecontentsource",
         "rpm.change_rpmdistribution",
         "rpm.change_rpmremote",
         "rpm.change_rpmrepository",
         "rpm.change_ulnremote",
         "rpm.delete_rpmalternatecontentsource",
         "rpm.delete_rpmdistribution",
         "rpm.delete_rpmpublication",
         "rpm.delete_rpmremote",
         "rpm.delete_rpmrepository",
         "rpm.delete_rpmrepository_version",
         "rpm.delete_ulnremote",
         "rpm.manage_roles_rpmalternatecontentsource",
         "rpm.manage_roles_rpmdistribution",
         "rpm.manage_roles_rpmpublication",
         "rpm.manage_roles_rpmremote",
         "rpm.manage_roles_rpmrepository",
         "rpm.manage_roles_ulnremote",
         "rpm.modify_content_rpmrepository",
         "rpm.refresh_rpmalternatecontentsource",
         "rpm.repair_rpmrepository",
         "rpm.sync_rpmrepository",
         "rpm.view_rpmalternatecontentsource",
         "rpm.view_rpmdistribution",
         "rpm.view_rpmpublication",
         "rpm.view_rpmremote",
         "rpm.view_rpmrepository",
         "rpm.view_ulnremote",
      ],
      "locked": true
    }

.. important::

>    Please note there are two types of Remotes with their permissions.

Reviewer: ```suggestion
   Please note there are two types of Remotes with their individual permissions.
```

RPM Viewer Role (rpm.viewer)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

A universal role allows you to view all content within the RPM plugin.

.. code-block:: shell
   :caption: Universal RPM viewer role

    {
      "pulp_href": "/pulp/api/v3/roles/d1e7bae0-2363-45e2-815c-d721ecc3133c/",
      "pulp_created": "2022-04-07T12:41:11.144883Z",
      "name": "rpm.viewer",
      "description": null,
      "permissions": [
        "rpm.view_rpmalternatecontentsource",
        "rpm.view_rpmdistribution",
        "rpm.view_rpmpublication",
        "rpm.view_rpmremote",
        "rpm.view_rpmrepository",
        "rpm.view_ulnremote",
      ],
      "locked": true
    }

Creating Roles
--------------

.. note::

    Roles shipped by the RPM plugin mentioned above are locked and cannot be modified or removed.

To create a new role, you need to specify its name and permissions to use.

> To find out which permissions you need for a new role, you can list an endpoints policies with the following command:

Reviewer: ```suggestion
To find out which permissions you need for a new role, you can list an endpoints policy with the following command:
```

.. code-block:: shell
   :caption: List an access policy of the repository endpoint

    pulp access-policy show --viewset-name "repositories/rpm/rpm"

**Example 1: RPM Repository Creator**

.. code-block:: shell
   :caption: Example of the RPM repository creator role

    pulp role create --name "rpm_repo_creator" \
        --permission "rpm.add_rpmrepository" \
        --permission "rpm.view_rpmrepository" \
        --permission "rpm.sync_rpmrepository" \
        --permission "rpm.modify_rpmrepository" \
        --permission "rpm.change_rpmrepository" \
        --permission "rpm.delete_rpmrepository" \
        --permission "rpm.add_rpmremote" \
        --permission "rpm.change_rpmremote" \
        --permission "rpm.delete_rpmremote" \
        --permission "rpm.view_rpmremote"

**Example 2: RPM Publisher**

.. code-block:: shell
   :caption: Example of the RPM publisher role

    pulp role create --name "rpm_publisher" \
        --permission "rpm.view_rpmrepository" \
        --permission "rpm.add_rpmpublication" \
        --permission "rpm.add_rpmdistribution"

.. code-block:: shell

> 

Reviewer: ```suggestion
   :caption: Globally assign these roles to users
   
```

    pulp user role-assignment add --username "bob" --role "rpm_repo_creator" --object ""
    pulp user role-assignment add --username "jack" --role "rpm_publisher" --object ""

.. warning::

   If a role contains multiple permissions over different objects, then you can't assign this role to a specific object.
   Trying to assign the role to a single object will fail.

>    For example, if a role contains ``add`` permission, you must use the empty ``--object`` param.

Reviewer: This is not true.
Having the `add` permission on an object does not lead anywhere, but it's possible.
Also assigning a role to an object is possible if at least one of it's permissions is applicable to that object.

Edit Access Policy
------------------

It is possible to edit the `access policy` for an endpoint.

.. code-block:: shell
   :caption: Change the access policy of the retrieve action without requiring any permission.

>     pulp access-policy update --viewset-name "remotes/rpm/rpm" --statements '[{"action": "retrieve", "effect": "allow"}]'

Reviewer: This is dangerous, as it deletes all but the retrieve action from this viewset.

Author: Fixed and reflect this information to the doc as well.

Role Based Access Control
=========================

.. important::

    This feature is a Tech Preview and may be subject of change.

.. important::

    User ``admin`` automatically gets all RBAC permissions.

RPM plugin presents Role Based Access for these endpoints :

* RPM Alternate Content Source (``/pulp/api/v3/acs/rpm/rpm/``)
* RPM Distributions (``/pulp/api/v3/distributions/rpm/rpm/``)
* RPM Publications (``/pulp/api/v3/publications/rpm/rpm/``)
* RPM Remotes (``/pulp/api/v3/remotes/rpm/rpm/`` and ``/pulp/api/v3/remotes/rpm/uln``)
* RPM Repository (``/pulp/api/v3/repositories/rpm/rpm/``)

Default Roles
-------------

RPM plugin ships a few default roles described bellow.

Viewer Role
^^^^^^^^^^^

> This role contains only ``view`` permission for an `endpoint`. Using this permission you can control who can

Reviewer 2: Not sure if `endpoint` needs to be a keyword

        "effect": "allow",
        "condition": [
            "has_model_perms:rpm.add_rpmpublication",
            "has_repo_attr_model_or_obj_perms:rpm.view_rpmrepository",
        ]
    }

Owner Role
^^^^^^^^^^

This role contains all permissions needed for manipulating objects, such as modifying or removing them.
This role also contains permission to manage role assignments on the object,
for example, adding or removing user roles.

By default, the user who created a new object will receive this role.

.. code-block:: shell
   :caption: Example of the RPM Remote owner role

    {
      "pulp_href": "/pulp/api/v3/roles/592356da-8cdf-4a1b-a434-3f10cda285f7/",
      "pulp_created": "2022-04-07T12:41:11.047168Z",
      "name": "rpm.rpmremote_owner",
      "description": null,
      "permissions": [
        "rpm.change_rpmremote",
        "rpm.delete_rpmremote",
        "rpm.manage_roles_rpmremote",
        "rpm.view_rpmremote"
      ],
      "locked": true
    }

RPM specific Roles
------------------

Also, the RPM plugin ships two roles to use.

RPM Admin Role (rpm.admin)
^^^^^^^^^^^^^^^^^^^^^^^^^^

This role is like a pulp admin but only for the RPM plugin.

> You can perform every RPM workflow. If you need a more

Reviewer 2: This role is like a pulp admin but only for the RPM plugin - it can perform any RPM workflow with no restrictions.

         "rpm.add_ulnremote",
         "rpm.change_rpmalternatecontentsource",
         "rpm.change_rpmdistribution",
         "rpm.change_rpmremote",
         "rpm.change_rpmrepository",
         "rpm.change_ulnremote",
         "rpm.delete_rpmalternatecontentsource",
         "rpm.delete_rpmdistribution",
         "rpm.delete_rpmpublication",
         "rpm.delete_rpmremote",
         "rpm.delete_rpmrepository",
         "rpm.delete_rpmrepository_version",
         "rpm.delete_ulnremote",
         "rpm.manage_roles_rpmalternatecontentsource",
         "rpm.manage_roles_rpmdistribution",
         "rpm.manage_roles_rpmpublication",
         "rpm.manage_roles_rpmremote",
         "rpm.manage_roles_rpmrepository",
         "rpm.manage_roles_ulnremote",
         "rpm.modify_content_rpmrepository",
         "rpm.refresh_rpmalternatecontentsource",
         "rpm.repair_rpmrepository",
         "rpm.sync_rpmrepository",
         "rpm.view_rpmalternatecontentsource",
         "rpm.view_rpmdistribution",
         "rpm.view_rpmpublication",
         "rpm.view_rpmremote",
         "rpm.view_rpmrepository",
         "rpm.view_ulnremote",
      ],
      "locked": true
    }

.. important::

>    Please note there are two types of Remotes with their individual permissions.

Author: I thought as well, but I found out that the access policy recognizes the class and fails when reusing the policy.

        --permission "rpm.change_rpmrepository" \
        --permission "rpm.delete_rpmrepository" \
        --permission "rpm.add_rpmremote" \
        --permission "rpm.change_rpmremote" \
        --permission "rpm.delete_rpmremote" \
        --permission "rpm.view_rpmremote"

**Example 2: RPM Publisher**

.. code-block:: shell
   :caption: Example of the RPM publisher role

    pulp role create --name "rpm_publisher" \
        --permission "rpm.view_rpmrepository" \
        --permission "rpm.add_rpmpublication" \
        --permission "rpm.add_rpmdistribution"

.. code-block:: shell
   :caption: Globally assign these roles to user

    pulp user role-assignment add --username "bob" --role "rpm_repo_creator" --object ""
    pulp user role-assignment add --username "jack" --role "rpm_publisher" --object ""

.. note::

   You can only assign roles at the object level if the role contains at least one permission applicable for that object.

Edit Access Policy
------------------

It is possible to edit the `access policy` for an endpoint.

.. code-block:: shell
   :caption: Change the access policy of all actions not requiring any permission.

    pulp access-policy update \

>         --viewset-name "remotes/rpm/rpm" \

Author: ``viewset-name`` is already in the CLI and already used in a few more plugins so changing it wouldn't be so easy.

>         --viewset-name "remotes/rpm/rpm" \

Reviewer 2: That is unfortunate.  @mdellweg Do you think this would be a worthwhile change, or do you think the cost would outweigh the benefit?  If we have to carry around the old commands indefinitely it's probably not worth it.

>         --viewset-name "remotes/rpm/rpm" \

Reviewer: The thing that i was told cannot change is the parameter in the REST api. The cli is just following the bad naming for consistency and less confusion and i'd rather keep it that way.

>         --viewset-name "remotes/rpm/rpm" \

Reviewer 2: +1

        --statements '[{"action": ["list", "retrieve"], "effect": "allow"}, {"action": "*", "effect": "allow"}]'

.. warning::

   When editing access policies, you must specify all the endpoint actions, or their usage will lead to failure.

You can modify ``creation-hooks`` to specify what will happen after object creation.
By default, a user will receive the ``owner`` role.

.. code-block:: shell
   :caption: Update creation hook of a repository to creator gets another role than `repository_owner`

>     pulp access-policy update --href "repositories/rpm/rpm" \

Author: ``--viewset-name`` is a better option in this place. ``--href`` is intended to use with the already created object.

>     pulp access-policy update --href "repositories/rpm/rpm" \

Reviewer: Yes href would not work...

Role Based Access Control
=========================

.. important::

>     This feature is a Tech Preview and can be subject of change.

Reviewer 2: This should be rephrased as "and *may* be subject *to* change"

I'm not really sure how to express the rule about "may" vs "can" but "may" is a little more formal and fits better in this situation.

.. important::

>     Using user ``admin`` will bypass all RBAC checks.

Reviewer 2: Does it actually bypass the checks or does the admin simply possess every permission?  I thought it was the latter https://github.com/pulp/pulp_rpm/pull/2418/files#diff-37aef788d3a12e65873c374ff05bc8629a8b636cab51ac31c127f94d43f1314eR234

The end result is the same but we should properly communicate what's going on :)

Author: rephrased to "admin gets all permissions"

Reviewer: The admin is assumed to have all the permissions (even before asking any permission backend). But the access policy can still disallow _any_ access. So there is actually a difference in the end result.

> * `Permission` : Permission to do **one** action on one `viewset` as a `view` or a `create`.

Reviewer 2: Should "as a" be ", for example" or "e.g."? I'm not sure what the intention is here

Author: This section should explain basic terms.

Author: But I would remove it.

>   * `Model Permission` : a permission for a model type, it means you can do this action on any object of the model

Reviewer 2: Are these docs meant to be for the user or for the plugin writer? If they're for a user, I would avoid using terms like "model" and "viewset".

    * E.g. add (create) which doesn't make sense for object if already exists.

  * `Object Permission` : a permission for `one specific object` only

    * E.g. update or delete. As such is applied for specific content.

* `Role` : Group of permission.

RPM plugin presents Role Based Access for these viewsets :

* RPM Alternate Content Source (``/pulp/api/v3/acs/rpm/rpm/``)
* RPM Distributions (``/pulp/api/v3/distributions/rpm/rpm/``)
* RPM Publications (``/pulp/api/v3/publications/rpm/rpm/``)
* RPM Remotes (``/pulp/api/v3/remotes/rpm/rpm/`` and ``/pulp/api/v3/remotes/rpm/uln``)

  * for both types there is one permission starting with ``rpm.rpmremote_[action]``

* RPM Repository (``/pulp/api/v3/repositories/rpm/rpm/``)

Default Roles
-------------

> RPM plugin ships few default roles described bellow.

Reviewer 2: "ships a few"

Viewer Role
^^^^^^^^^^^

> This role contains only ``view`` permission for a `viewset`. Using this permission you can control who can

Reviewer 2: As with above, most users won't know what a viewset is.

view specific content for example a repository.

.. code-block:: shell
   :caption: You can check the role by pulp-cli

    pulp role show --name "rpm.rpmremote_viewer"

.. code-block:: shell

    {
      "pulp_href": "/pulp/api/v3/roles/62baf888-09c2-4c18-8d3a-354d9c6bdb48/",
      "pulp_created": "2022-04-07T12:41:11.065567Z",
      "name": "rpm.rpmremote_viewer",
      "description": null,
      "permissions": [
        "rpm.view_rpmremote"
      ],
      "locked": true
    }

Creator Role
^^^^^^^^^^^^

This role contain only ``add`` permission for a specific `viewset` to allow user create a new object of its type.

> Be aware that it is possible you will need additional permissions to be able create new object.

Reviewer 2: "to be able create new object" should be switched to:

"to be able to create a new object" or "to be able to create new objects"

E.g. ``rpm.rpmrepository_view`` for a repository object to create a new publication using this repository.

.. code-block:: shell

>    :caption: Example of requested condition to create a RPM Publication

Reviewer 2: of the

    {
        "action": ["create"],
        "principal": "authenticated",
        "effect": "allow",
        "condition": [
            "has_model_perms:rpm.add_rpmpublication",
            "has_repo_attr_model_or_obj_perms:rpm.view_rpmrepository",
        ]
    }

Owner Role
^^^^^^^^^^

> This role contains all permissions together that are needed for manipulating object such as change (if possible) or

Reviewer 2: "manipulating object such as change (if possible) or remove it" should be changed to

"manipulating objects, such as modifying or removing it"

> remove it. This role also contains permission to manage permission on the object so you can add or remove permissions

Reviewer 2: "This role also contains permission to manage permission on the object so you can add or remove permissions" should be changed to

"This role also contains the permission to manage permissions on the object, e.g. adding or removing permissions"

for different users or groups.

> By default this role is assigned to user who created new object.

Reviewer 2: "this role is assigned to the user who created the new object"

.. code-block:: shell

>    :caption: Example of RPM Remote owner role

Reviewer 2: Example of the (or "using the"?)

    {
      "pulp_href": "/pulp/api/v3/roles/592356da-8cdf-4a1b-a434-3f10cda285f7/",
      "pulp_created": "2022-04-07T12:41:11.047168Z",
      "name": "rpm.rpmremote_owner",
      "description": null,
      "permissions": [
        "rpm.change_rpmremote",
        "rpm.delete_rpmremote",
        "rpm.manage_roles_rpmremote",
        "rpm.view_rpmremote"
      ],
      "locked": true
    }

RPM specific Roles
------------------

> Also, RPM plugin ships two basic roles which you can use.

Reviewer 2: Also, the RPM plugin

> Role rpm.admin

Reviewer 2: "Admin Role (rpm.admin)"

        "rpm.add_rpmalternatecontentsource",
        "rpm.change_rpmalternatecontentsource",
        "rpm.delete_rpmalternatecontentsource",
        "rpm.manage_roles_rpmalternatecontentsource",
        "rpm.refresh_rpmalternatecontentsource",
        "rpm.view_rpmalternatecontentsource",
        "rpm.add_rpmdistribution",
        "rpm.change_rpmdistribution",
        "rpm.delete_rpmdistribution",
        "rpm.manage_roles_rpmdistribution",
        "rpm.view_rpmdistribution",
        "rpm.add_rpmpublication",
        "rpm.delete_rpmpublication",
        "rpm.manage_roles_rpmpublication",
        "rpm.view_rpmpublication",
        "rpm.add_rpmremote",
        "rpm.change_rpmremote",
        "rpm.delete_rpmremote",
        "rpm.manage_roles_rpmremote",
        "rpm.view_rpmremote",
        "rpm.add_rpmrepository",
        "rpm.change_rpmrepository",
        "rpm.copy_rpmrepository",
        "rpm.delete_rpmrepository",
        "rpm.manage_roles_rpmrepository",
        "rpm.modify_rpmrepository",
        "rpm.repair_rpmrepository",
        "rpm.sync_rpmrepository",
        "rpm.view_rpmrepository"
      ],
      "locked": true
    }

Role rpm.viewer
^^^^^^^^^^^^^^^

> This role is universal viewer who can view all guarded endpoints within RPM plugin.

Reviewer 2: within the RPM plugin

.. code-block:: shell
   :caption: Universal RPM viewer role

    {
      "pulp_href": "/pulp/api/v3/roles/d1e7bae0-2363-45e2-815c-d721ecc3133c/",
      "pulp_created": "2022-04-07T12:41:11.144883Z",
      "name": "rpm.viewer",
      "description": null,
      "permissions": [
        "rpm.view_rpmalternatecontentsource",
        "rpm.view_rpmdistribution",
        "rpm.view_rpmpublication",
        "rpm.view_rpmremote",
        "rpm.view_rpmrepository"
      ],
      "locked": true
    }

Creating Roles
--------------

.. note::

>     Roles shipped by RPM plugin mentioned above are locked and cannot be edited or removed.

Reviewer 2: by the RPM plugin

To create a role it is enough to specify its name and which permissions to use.
To find out which permissions are needed for a new role, access policy of viewset can be listed with following command:

.. code-block:: shell
   :caption: List access policy

    pulp access-policy show --viewset-name "repositories/rpm/rpm"

**Example 1: RPM Repository Creator**

.. code-block:: shell
   :caption: Example of RPM repository creator

    pulp role create --name "rpm_repo_creator" \
        --permission "rpm.add_rpmrepository" \
        --permission "rpm.view_rpmrepository" \
        --permission "rpm.sync_rpmrepository" \
        --permission "rpm.modify_rpmrepository" \
        --permission "rpm.change_rpmrepository" \
        --permission "rpm.delete_rpmrepository" \
        --permission "rpm.add_rpmremote" \
        --permission "rpm.change_rpmremote" \
        --permission "rpm.delete_rpmremote" \
        --permission "rpm.view_rpmremote"

**Example 2: RPM Publisher**

.. code-block:: shell

>    :caption: Example of RPM publisher

Reviewer 2: Example of an RPM Publisher role

Role Based Access Control
=========================

.. important::

    This feature is a Tech Preview and may be subject of change.

.. important::

    User ``admin`` automatically gets all RBAC permissions.

> RPM plugin presents Role Based Access for these endpoints :

Reviewer 3: what about copy, compsxml and rest of the content?

Author: Added note about the content and rpm copy.
